lib/msal-common/src/client/BrokerAuthorizationCodeClient.ts

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { AuthorizationCodeClient } from "./AuthorizationCodeClient";
+import { AuthorizationUrlRequest } from "../request/AuthorizationUrlRequest";
+import { RequestParameterBuilder } from "../request/RequestParameterBuilder";
+import { AuthorizationCodeRequest } from "../request/AuthorizationCodeRequest";
+import { StringUtils } from "../utils/StringUtils";
+import { ClientAuthError } from "../error/ClientAuthError";
+import { ResponseHandler } from "../response/ResponseHandler";
+import { BrokerAuthenticationResult } from "../response/BrokerAuthenticationResult";
+
+/**
+ * Oauth2.0 Authorization Code client implementing the broker protocol for browsers.
+ */
+export class BrokerAuthorizationCodeClient extends AuthorizationCodeClient {
+
+    /**
+     * API to acquire a token in exchange of 'authorization_code` acquired by the user in the first leg of the
+     * authorization_code_grant
+     * @param request
+     */
+    async acquireToken(request: AuthorizationCodeRequest, cachedNonce?: string, cachedState?: string): Promise<BrokerAuthenticationResult> {
+        this.logger.info("in acquireToken call");
+        // If no code response is given, we cannot acquire a token.
+        if (!request || StringUtils.isEmpty(request.code)) {
+            throw ClientAuthError.createTokenRequestCannotBeMadeError();
+        }
+
+        const response = await this.executeTokenRequest(this.authority, request);
+
+        const responseHandler = new ResponseHandler(
+            this.config.authOptions.clientId,
+            this.cacheManager,
+            this.cryptoUtils,
+            this.logger
+        );
+
+        // Validate response. This function throws a server error if an error is returned by the server.
+        responseHandler.validateTokenResponse(response.body);
+        const tokenResponse = responseHandler.handleBrokeredServerTokenResponse(response.body, this.authority, cachedNonce, cachedState);
+
+        return tokenResponse;
+    }
+
+    /**
+     * This API validates the `AuthorizationCodeUrlRequest` and creates a URL
+     * @param request
+     */
+    protected createAuthCodeUrlQueryString(request: AuthorizationUrlRequest): string {
```

Comment:
Why override the existing function? Can't we just add the params we need to `AuthorizationUrlRequest` and add them to the existing function on `AuthorizationCodeClient`?

Reply:
Subclassing and overriding allows us to keep the broker code separate from the general authorization code flow. This way we don't affect msal-node flows either.

Reply:
After some more consideration I've come around to this. I think this approach makes sense.